components/lib/push-amazon/src/main/java/mozilla/components/lib/push/amazon/AbstractAmazonPushService.kt

```diff
@@ -0,0 +1,74 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ *  License, v. 2.0. If a copy of the MPL was not distributed with this
+ *  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package mozilla.components.lib.push.amazon
+
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import com.amazon.device.messaging.ADM
+import com.amazon.device.messaging.ADMMessageHandlerBase
+import mozilla.components.concept.push.EncryptedPushMessage
+import mozilla.components.concept.push.PushError
+import mozilla.components.concept.push.PushProcessor
+import mozilla.components.concept.push.PushService
+import androidx.annotation.VisibleForTesting
+
+/**
+ * An Amazon Cloud Messaging implementation of the [PushService] for Android devices that support Google Play Services.
+ * [ADMMessageHandlerBase] requires a redundant constructor parameter.
+ */
+abstract class AbstractAmazonPushService : ADMMessageHandlerBase("AbstractAmazonPushService"), PushService {
+    private lateinit var adm: ADM
+
+    override fun start(context: Context) {
+        adm = ADM(context)
+        if (adm.registrationId == null) {
+            adm.startRegister()
+        }
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
+    public override fun onRegistered(registrationId: String) {
+        PushProcessor.requireInstance.onNewToken(registrationId)
+    }
+
+    override fun onRegistrationError(errorId: String) {
+        PushError.Registration("registration failed: $errorId")
+    }
+
+    override fun onUnregistered(registrationId: String) {
+        PushError.Registration("account no longer associated with device")
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
```

Comment:
Do we need these annotations for public overrides?

Reply:
It's protected fun by super, abstracts are intended to extended; only making it public for mocking test visibility. We could leave it public of course, of course, what do you think?

Reply:
Ah gotcha, no this is fine then. Thanks!